### PR TITLE
[openshift/deploy.sh] delete previously allocated AWS resources

### PR DESCRIPTION
**Pull request by jpopelka**
Please review with care as this is potentially destroying functionality.

1.  Previously if we found already existing RDS database, we were recreating it.
With this change we by default purge also previously allocated (`$DEPLOYMENT_PREFIX` prefixed) SQS queues, S3 buckets, DynamoDB tables.
The purging can be avoided by using `--keep-resources` argument.
The question is whether we by default really want to purge all those AWS resources, or whether we by default want to keep some of them (like S3 buckets?) ?

2. The new `hack/purge_AWS_resources.py` has also a `-t/--tag` option, which when specified, causes the script to delete all AWS resources [tagged](https://github.com/openshiftio/openshift.io/issues/1081) with this tag.
Fixes https://github.com/openshiftio/openshift.io/issues/1330

### CODE REVIEW

**Comment by jpopelka** [comment]
[test]

**Comment by jpopelka** [comment]
[test]

File: openshift/deploy.sh [hunk 1]
```diff
@@ -159,13 +182,28 @@ function wait_for_rds_instance_info() {
     done
 }
 
+clean_resources=true # default
+while [[ $# -gt 0 ]]; do
+    key="$1"
+    case $key in
+        --keep-resources)
+            clean_resources=false
+            shift # past argument
```
Comment (tisnik) [inline openshift/deploy.sh]: (minor) should not it be next argument?

**Comment by tisnik** [comment]
LGTM, I'd just add test if DEPLOYMENT_PREFIX is not empty (not sure what happened if it is, probably nothing catastrophic though)

**Comment by jpopelka** [comment]
> I'd just add test if DEPLOYMENT_PREFIX is not empty

thanks, already fixed in to-be-pushed commit

**Comment by jpopelka** [comment]
Ready for review, anybody feel free to review/test/comment.

**Comment by jpopelka** [comment]
@tuxdna if you're playing with the `deploy.sh` script, you might take a look at this PR too ;-) thanks

File: openshift/deploy.sh [hunk 2]
```diff
@@ -37,7 +41,22 @@ fabric8-analytics-worker fabric8-analytics-pgbouncer gremlin-docker
 fabric8-analytics-scaler fabric8-analytics-firehose-fetcher
 fabric8-analytics-license-analysis fabric8-analytics-stack-analysis fabric8-analytics-stack-report-ui"
 
+clean_resources=true # default
+while [[ $# -gt 0 ]]; do
+    key="$1"
```
Comment (tisnik) [inline openshift/deploy.sh]: could not we use:
```
for key in "$@"; do
```

File: openshift/purge_AWS_resources.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    """Operations to purge/delete previously allocated AWS resources."""
+
+    def __init__(self, tag):
+        """Initialize object."""
+        self.tag = tag
+
+    def purge_sqs_queues(self):
+        """Purge/Delete SQS queues.
+
+        If tag was specified during object creating, then delete all queues tagged with this tag.
```
Comment (tisnik) [inline openshift/purge_AWS_resources.py]: during object creati**on**?

File: openshift/purge_AWS_resources.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        except ValueError:
+            print("Tag needs to be 'key,value'")
+            exit(1)
+        print("About to delete resources tagged with %r" % ','.join(args.tag))
+    else:
+        print("About to purge %r prefixed resources." % _DEPLOYMENT_PREFIX)
+
+    if (args.tag and args.tag[1].lower() in {'prod', 'stage'}
+            or _DEPLOYMENT_PREFIX.lower() in {'prod', 'stage'}) and not args.force:
+        print("If you really want to destroy prod/stage use -f/--force.")
```
Comment (tisnik) [inline openshift/purge_AWS_resources.py]: 🥇

Comment (tisnik) [reply openshift/purge_AWS_resources.py]: "If you really want to destroy {environment} use -f/--force".format(environment=args.tag[1] ?

**Comment by tisnik** [comment]
@jpopelka LGTM, but it would be nice to know @msrb opinion as well, as this script can do some not-so-safe things ;)

File: openshift/deploy.sh (hunk 2 above)
Comment (jpopelka) [reply openshift/deploy.sh]: Sure we can, changed.

**Comment by tuxdna** [comment]
I would keep the non-destructive behavior as default, and whenever the deploy script is run print a message saying 

> If you want to delete / recreate AWS resources viz RDS, S3, SQS etc, then run with `--keep-resources=false`

**Comment by jpopelka** [comment]
@tuxdna I reverted the default behaviour per your comment. See latest commit.

**Comment by tuxdna** [comment]
The latest changes look good to me. I do have one comment below:

If I now provide `--purge-aws-resources` argument, I can see that this will cause purge of S3, SQS and Dynamodb data ( via `python3 ./purge_AWS_resources.py` ). However I can't figure out the flow for purging RDS data. In case the requred RDS database already exists, and I have provided `--purge-aws-resources` flag, will the RDS database be re-used or will it be first deleted and then recreated again ?

**Comment by tuxdna** [comment]
Okay, after running `deploy.sh --purge-aws-resources` against devcluster, I figured that database is dropped first and created:

```
DB instance myprefix-bayesiandb already exists
Trying to get RDS DB endpoint for myprefix-bayesiandb ...
recreating database
DROP DATABASE
CREATE DATABASE
Tagging RDS instance with Key=ENV,Value=myprefix

```

I guess this will happen even when we don't provide `--purge-aws-resources` and the database already exists. Is that correct understanding ?

**Comment by tuxdna** [comment]
[test]

**Comment by tuxdna** [comment]
This is the output I get from running deploy.sh script.

**Comment by tuxdna** [comment]
Tests are failing for this PR, which I believe should be unrelated to changes in this PR.

```
Failing scenarios:

  features/smoketest.feature:4  Check the API entry point

  features/smoketest.feature:10  Check the /schemas entry point

  features/smoketest.feature:16  Check the /system/version entry point

  features/smoketest.feature:22  Check the jobs API entry point

  features/server_api.feature:4  Check the API entry point

  features/server_api.feature:11  Check the /api/v1/readiness response

  features/server_api.feature:18  Check the /api/v1/liveness response

  features/server_api.feature:26  Check the service/state response

  features/server_api.feature:34  Check the /api/v1/schemas response

  features/component_search.feature:3  Check the component search functionality for nonexistent component from npm ecosystem

  features/stack_analyses_v2_minimal.feature:43  Check that the stack-analyses returns a valid response for python ecosystem

  features/api_backbone.feature:3  Check the Backbone API /api/v1/stack_aggregator response

  features/api_backbone.feature:10  Check the Backbone API /api/v1/recommender response

```

Could you try merging with master and then test ?

**Comment by jpopelka** [comment]
> Okay, after running deploy.sh --purge-aws-resources against devcluster, I figured that database is dropped first and created.
> I guess this will happen even when we don't provide --purge-aws-resources and the database already exists. Is that correct understanding ?

Yes, that's correct. I wanted to keep the default behaviour unchanged, i.e. clean RDS, but keep other resources untouched.
It's also described in [README](https://github.com/jpopelka/fabric8-analytics-common/blob/659f823a99fd92753b0aac6b53d19eb1212674cf/openshift/README.md#deploy-fabric8-analytics-services).

**Comment by jpopelka** [comment]
[test]

**Comment by msrb** [comment]
Yeah, I agree that we want to either purge all or none. But I don't have strong opinion on which of those options should be the default one. We certainly want to make sure that we never purge any `prod-` or `-STAGE/stage` resources during deployment. This is strictly for dev cluster deployments.

**Comment by jpopelka** [comment]
> Yeah, I agree that we want to either purge all or none.

OK, amended latest commit.

> But I don't have strong opinion on which of those options should be the default one.

For now, default is the safer, non-purging one.

> We certainly want to make sure that we never purge any prod- or -STAGE/stage resources during deployment. This is strictly for dev cluster deployments.

Yes, there's a [safety check](https://github.com/fabric8-analytics/fabric8-analytics-common/pull/446/files#diff-ea692428daf4334ddb711bd2ec1a7f94R164)

**Comment by jpopelka** [comment]
[test]

**Comment by tuxdna** [comment]
One test is still failing

```
Failing scenarios:

  features/stack_analyses_v2_minimal.feature:43  Check that the stack-analyses returns a valid response for python ecosystem

```

**Comment by msrb** [comment]
I don't think deployment scripts are covered by the test suite. I think we should decouple the scripts from integration tests. This repository is kinda mess.

**Comment by tuxdna** [comment]
@msrb Yeah makes sense. I doubt if the deploy scripts affect the integration tests in any way.

**Comment by tuxdna** [comment]
[test]

**Comment by jpopelka** [comment]
[test]

**Comment by humaton** [comment]
LGTM 
There is one test failing but it's not related to this work.
